java-checks/src/main/java/org/sonar/java/checks/SillyEqualsCheck.java

```diff
@@ -0,0 +1,124 @@
+  key = "S2159",
+  name = "Silly equality checks should not be made",
+  tags = {"bug", "unused"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class SillyEqualsCheck extends SubscriptionBaseVisitor {
+
+  private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+  private static final String MESSAGE = "Remove this call to \"equals\"; comparisons between unrelated types always return false.";
+
+  private static final MethodInvocationMatcher EQUALS_METHOD = MethodInvocationMatcher.create()
+    .name("equals")
+    .addParameter(JAVA_LANG_OBJECT);
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree methodTree = (MethodInvocationTree) tree;
+    if (EQUALS_METHOD.matches(methodTree)) {
+      ExpressionTree firstArgument = Iterables.getOnlyElement(methodTree.arguments());
+      Type argumentType = firstArgument.symbolType().erasure();
+      Type ownerType = getMethodOwnerType(methodTree).erasure();
+      if (isLiteralNull(firstArgument)) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons against null always return false; consider using '== null' to check for nullity.");
+      } else if (ownerType.isArray()) {
+        checkWhenOwnerIsArray(tree, ownerType, argumentType);
+      } else {
+        checkWhenOwnerIsNotArray(tree, ownerType, argumentType);
+      }
+    }
+  }
+
+  private void checkWhenOwnerIsArray(Tree tree, Type ownerType, Type argumentType) {
+    if (!argumentType.isArray()) {
+      if (!argumentType.is(JAVA_LANG_OBJECT)) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons between an array and a type always return false.");
+      }
+    } else if (!areRelated(((Type.ArrayType) ownerType).elementType(), ((Type.ArrayType) argumentType).elementType())) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between unrelated arrays always return false.");
+    } else {
+      addIssue(tree, "Use \"Arrays.equals(array1, array2)\" or the \"==\" operator instead of using the \"Object.equals(Object obj)\" method.");
+    }
+  }
+
+  private void checkWhenOwnerIsNotArray(Tree tree, Type ownerType, Type argumentType) {
+    if (argumentType.isArray() && !ownerType.is(JAVA_LANG_OBJECT)) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between a type and an array always return false.");
+    } else if (argumentType.isClass() && !areRelated(ownerType, argumentType)) {
+      if (isFinalClassWithInterface(ownerType, argumentType)) {
+        addIssue(tree, MESSAGE);
+      } else if (!ownerType.symbol().isInterface() && !argumentType.symbol().isInterface()) {
+        addIssue(tree, MESSAGE);
+      }
+    }
+  }
+
+  private boolean isFinalClassWithInterface(Type type1, Type type2) {
```

Comment:
name of method and parameters is not really explicit : 
callEqualsOnAFinalClassWithInterfaceArgument(callee, argument);